### DISCUSSION TITLE
FR: keybindings to jump quickly to code diffs in magit-revision buffers

### DISCUSSION THREAD

**Discussion by aspiers** [magit/magit#4884] [Ideas]
I find myself regularly missing the ability to jump straight to the first diff hunk in a magit-revision buffer.

In 2014 (wow, how is it 9 years ago already!) I proposed the keybinding `j` in #1574 for jumping from lines in the diffstats section to the corresponding file below, which @tarsius very kindly implemented as `magit-jump-to-diffstat-or-diff`, and I have found it incredibly useful ever since.

However:

- if I want to jump from the top of the buffer to the first diff file, this still requires the sequence: `j` `n` `j`, or
- if I want to jump from the top of the buffer to the first diff hunk, this still requires the sequence: `j` `n` `j` `n`.

It would be nice to be able to achieve either of these in two keystrokes or less.  Possible ways of achieving this:

1. Change `magit-jump-to-diffstat-or-diff` to jump one line lower than it does currently (at least, if the diffstats section is expanded).  This would reduce the above two jumps to `j` and `j` `j` `n`; however that's still 3 keystrokes for the latter.

2. Change `magit-jump-to-diffstat-or-diff` so that if it's re-executed when on the header of the diffstats section, it jumps to the header of the first file.  This would reduce the above two jumps to `j` `j` and `j` `j` `n`; however that's still 3 keystrokes for the latter.

3. Change `magit-jump-to-diffstat-or-diff` so that if it's executed above the header of the diffstats section, it jumps to the header of the first diff hunk of the first file.

   Additionally change it so that if it's executed anywhere on or within a hunk, it jumps back to the corresponding file within the diffstats section.  (Currently in this case it jumps back to the *header* of the diffstats section which is not as useful, especially if the revision changes many files, so actually I suggest making this second change regardless.)

   This would reduce the above two jumps to `j` `j` and `j` respectively, which is optimal and also yields improved behaviour in other cases as explained in the previous sentence.

The third of these 3 options seems the best to me: it covers all cases in the minimum number of keystrokes, whilst adding useful new behaviour.

**Comment by tarsius** [comment]
Thanks for the suggestions!

I am currently focused on bug slaying and getting ready for releases. So it likely it will be a while until I get around looking at feature requests such as this.

**Reply by aspiers** [reply]
No problem at all, thanks a lot for letting me know!

**Reply by tarsius** [reply]
Since v4.4.0 `j` is bound to `magit-revision-jump`, which requires fewer key presses. Jumping from the top to the first file is `j d`, for example.
